akka-stream/src/main/scala/akka/stream/impl/Sinks.scala

```diff
@@ -548,3 +545,210 @@ final private[stream] class LazySink[T, M](sinkFactory: T ⇒ Future[Sink[T, M]]
     (stageLogic, promise.future)
   }
 }
+
+/**
+ * INTERNAL API
+ */
+final class FoldResourceSink[T, S](
+  open:      () ⇒ S,
+  writeData: (S, T) ⇒ Unit,
```

Comment:
Thinking out aloud: Should this be `(S, T) => S` to allow a more functional style where each write may return a new `S` (and imperative code can just return the `S` as is, or maybe even `(S, T) => Option[S]` to allow the sink to stop early by returning a `None`, ofc these two does not align, because then there would be no `S` for the closing logic?